RHIDP-5422: TechDocs add-ons doc (#975)
redhat-developer/red-hat-developers-documentation-rhdh: modules/techdocs/proc-techdocs-addon-create.adoc

. In the `index.ts` file, export the custom TechDocs add-on that you want to create by adding the following code:
+
[source,java,subs="+attributes,+quotes"]
----
export { _<new_plugin_for_techdocs_add-on>_, _<new_techdocs_add-on>_ } from './plugin';
----
. Create a new `_<new_techdocs_add-on>_.tsx` file and add the code for your new TechDocs add-on component.
+
////
[source,java,subs="+attributes,+quotes"]
----
can add example code for this file, if helpful
can also mention a template that the user can configure, if there is one
----
////
. Create a new `index.tsx` file and use it to export your new TechDocs add-on component by adding the following code:
+
[source,java,subs="+attributes,+quotes"]
----
export { _<new_techdocs_add-on>_, type _<new_techdocs_addon_props>_} from './_<new_techdocs_add-on_directory>_'
----
+
where

_<new_techdocs_addon_props>_ (Optional) :: Specifies the `props` for your new TechDocs add-on, as specified in your `_<new_techdocs_add-on>_.tsx` file, if applicable.
. In the `plugins.ts` file, import your new TechDocs add-on component.

> . Install and configure your new TechDocs add-on by following the steps in link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.5/html/techdocs_for_red_hat_developer_hub/assembly-techdocs-addons-installing.adoc[Installing and configuring a TechDocs add-on].

Reviewer: ```suggestion
. Install and configure your new TechDocs add-on by following the steps in link:https://docs.redhat.com/en/documentation/red_hat_developer_hub/{product-version}/html/techdocs_for_red_hat_developer_hub/assembly-techdocs-addons-installing.adoc[Installing and configuring a TechDocs add-on].
```
Use an attribute, so we do not have to update this link for every release. In other products, the link title and URL are also defined using attributes.

              _<new_techdocs_add-on_fetch-component>_.test.tsx
              _<new_techdocs_add-on_fetch-component>_.tsx
               index.ts
       index.ts
       plugin.test.ts
       plugin.ts
       routes.ts
       setupTests.ts
   .eslintrc.js
   package.json
   README.md
----

.Prerequisites
* The `yarn` package manager is installed.
* Docker v3.2.0 or later or Podman v3.2.0 or later is installed and running.

.Procedure
. In the terminal, navigate to the root folder of the repository where you want to create your new plugin.
. To create a new front-end plugin, run the following command:
+
[source,terminal,subs="+attributes,+quotes"]
----
yarn new
----
.Example output
+
[source,terminal,subs="+quotes"]
----
? What do you want to create? plugin - A new frontend plugin
? Enter the ID of the plugin [required]
----
+
. In the terminal prompt, type a name for the new plugin. For example:
+
[source,terminal,subs="+attributes,+quotes"]
----
? Enter the ID of the plugin [required] _<new_plugin_for_techdocs_add-on>_
----
+
.Example output
+
[source,terminal,subs="+attributes,+quotes"]
----
Successfully created plugin
----
+
[NOTE]
====

> Running the `yarn new` command automatically generates  a sub-directory with

Reviewer 2: ```suggestion
Running the `yarn new` command automatically generates a sub-directory with
```

    component: _<new_techdocs_add-on_component>_,
  }),
);
----
+
where

_<new_plugin_for_techdocs_add-on>_ :: Specifies the new plugin that you use to import the TechDocs add-on to your {product} instance.
_<new_techdocs_add-on>_ :: Specifies the custom TechDocs add-on that you want to create.
_<new_techdocs_addon_props>_ (Optional) :: Specifies the `props` for your new TechDocs add-on, as specified in your `_<new_techdocs_add-on>_.tsx` file, if applicable.
. In the `plugins.ts` file, import you new TechDocs add-on component.
_<new_techdocs_add-on_component>_ :: Specifies the React component for the custom TechDocs add-on that you want to create. You will create this component in a `.tsx` file in a later step.
. In the `index.ts` file, export the custom TechDocs add-on that you want to create by adding the following code:
+
[source,java,subs="+attributes,+quotes"]
----
export { _<new_plugin_for_techdocs_add-on>_, _<new_techdocs_add-on>_ } from './plugin';
----
. Create a new `_<new_techdocs_add-on>_.tsx` file and add the code for your new TechDocs add-on component.
+
////
[source,java,subs="+attributes,+quotes"]
----
can add example code for this file, if helpful

> can also mention a template that the user can configure, if there is one

Reviewer 2: This is a TODO?

Reviewer 2: @linfraze Example code could be a link to code of existing backstage techdocs addons (I think you have mentioned we can't link to them unfortunately)? I can create a separate repo for example addon I created before, but it is unofficial and not tested :).
I don't think there is a template for creating an addon.

----
////
. Create a new `index.tsx` file and use it to export your new TechDocs add-on component by adding the following code:
+
[source,java,subs="+attributes,+quotes"]
----
export { _<new_techdocs_add-on>_, type _<new_techdocs_addon_props>_} from './_<new_techdocs_add-on_directory>_'
----
+
where

_<new_techdocs_addon_props>_ (Optional) :: Specifies the `props` for your new TechDocs add-on, as specified in your `_<new_techdocs_add-on>_.tsx` file, if applicable.
. In the `plugins.ts` file, import you new TechDocs add-on component.

> . Navigate to `packages/app/src/components/catalog/EntityPage.tsx` and your new TechDocs add-on. For example:
> +
> [source,java,subs="+attributes,+quotes"]
> ----
> ...
> const techdocsContent = (
>   <EntityTechdocsContent>
>     <TechDocsAddons>
>       <ReportIssue />
>       <_<new_techdocs_add-on>_ />
>     </TechDocsAddons>
>   </EntityTechdocsContent>
> );
> ...

Reviewer 2: This one is not needed for RHDH.
```suggestion
```

----

.Verification

> . In the terminal, navigate to the `rbac` directory.

Reviewer 2: I would use this as verification:

Verification

- Restart the RHDH application and verify that the plugin is successfully activated and configured.
- Verify the application logs for confirmation and ensure the plugin is functioning as expected.

// Module included in the following assemblies:
//
// [WIP] file created but not currently part of any assembly or title

:_mod-docs-content-type: PROCEDURE
[id="proc-techdocs-addon-create_{context}"]
== Creating a TechDocs add-on

If your organization has documentation needs that are not met by the functions of existing TechDocs add-ons, developers can create a new add-on for your TechDocs plugin.

A TechDocs add-on is a react component that is imported from a front-end plugin. If you do not have an existing plugin for your TechDocs add-on, you can create a new plugin by configuring the default front-end plugin files in the `backstage/community-plugins` repository. After you create a new front-end plugin and a new TechDocs add-on, you can use the new plugin to import the new TechDocs add-on into your {product} instance.

> The folder structure of a new plugin looks similar to the following example:

Reviewer 2: This is after adjusting the code. So maybe: The folder structure of a new frontend plugin with exported TechDocs addon looks similar to the following example:

Reviewer 2: Or we can have here only the default generated structure listed and then in the section where we explain how to create the addon we can show the adjusted structure.

Author: Made a change. Please check for accuracy.

[source,json,subs="+attributes,+quotes"]
----
_<new_plugin_for_techdocs_add-on>_/
   dev/
       index.ts

>    node_modules/

Reviewer 2: ```suggestion
```

// Module included in the following assemblies:
//

> // [WIP] file created but not currently part of any assembly or title

Author: SME - writer can benefit from a live or recorded demo on how to create a custom dynamic plugin (add-on)

Author: OCP docs here: https://docs.openshift.com/container-platform/4.18/web_console/dynamic-plugin/dynamic-plugins-get-started.html#dynamic-plugin-development_dynamic-plugins-get-started 
Backstage docs here: https://backstage.io/docs/features/techdocs/addons#creating-an-addon

Author: @dzemanov it looks like the link you sent covers the use case of _installing third-party add-ons_, which is documented by `proc-techdocs-addon-install-third-party.adoc`, but I expected the procedure for _creating/developing a custom add-on_ to be different than installing an existing third-party add-on. 

If creating a custom add-on is no different than installing an existing third-party add-on, then we can delete `proc-techdocs-addon-create.adoc` and consider that use case covered. If they are different processes, then the docs team will need to understand how to create a custom add-on so we can document that procedure.

Reviewer 2: Sent a new link with creating example techdocs addon.

[id="proc-techdocs-addon-create_{context}"]
== Creating a TechDocs add-on

If your organization has documentation needs that are not met by the functions of existing TechDocs add-ons, developers can create a new add-on for your TechDocs plugin.

> A TechDocs add-on is a react component that is imported from a front-end plugin. If you do not have an existing plugin for your TechDocs add-on, you can create a new plugin by configuring the default front-end plugin files in the `backstage/community-plugins` repository. After you create a new front-end plugin and a new TechDocs add-on, you can use the new plugin to import the new TechDocs add-on into your {product} instance.

Reviewer 2: I totally missed this section, pardon! I am reading the deployed version so I can see if something is formatted incorrectly. Should this section be added to the main navigation for techdocs addons? Should it be rendered somewhere after 3.3. Using TechDocs add-ons?

> A TechDocs add-on is a react component that is imported from a front-end plugin. If you do not have an existing plugin for your TechDocs add-on, you can create a new plugin by configuring the default front-end plugin files in the `backstage/community-plugins` repository. After you create a new front-end plugin and a new TechDocs add-on, you can use the new plugin to import the new TechDocs add-on into your {product} instance.

Author: Yes! Good catch. At one point I wasn't sure if we were actually going to include this use case, so I do need to `include` this after the _Using_ section

The folder structure of a new plugin looks similar to the following example:
+
[source,json,subs="+attributes,+quotes"]
----
_<new_plugin_for_techdocs_add-on>_/
   dev/
       index.ts
   node_modules/
   src/
       components/
         _<new_techdocs_add-on_component>_/
              _<new_techdocs_add-on_component>_.test.tsx
              _<new_techdocs_add-on_component>_.tsx
               index.ts
         _<new_techdocs_add-on_fetch-component>_/
              _<new_techdocs_add-on_fetch-component>_.test.tsx
              _<new_techdocs_add-on_fetch-component>_.tsx
               index.ts
       index.ts
       plugin.test.ts
       plugin.ts
       routes.ts
       setupTests.ts
   .eslintrc.js
   package.json
   README.md
----

> .Prerequisites

Author: @Gerry-Forde I like that approach

* The `yarn` package manager is installed.
* Docker v3.2.0 or later or Podman v3.2.0 or later is installed and running.

.Procedure

> . In the terminal, navigate to the root folder of your {product} instance.

Reviewer 2: He can create one running `npx @backstage/create-app@latest`
